### PR TITLE
(PDB-1810) Add code_id to catalogs

### PR DESCRIPTION
**Pull request by ajroetker**
This commit adds the Direct Puppet related code_id to catalogs for APPMGMT.

### CODE REVIEW

**Comment by wkalt** [comment]
@puppetlabs-jenkins test this please

File: documentation/api/command/v1/commands.markdown [hunk 1]
```diff
@@ -96,7 +97,7 @@ processed.
 * The catalog 'name' field has been renamed to 'certname'.
 
 The payload is expected to be a Puppet catalog, as a JSON object,
-conforming exactly to the [catalog wire format v6][catalogv6]. Extra
+conforming exactly to the [catalog wire format v7][catalogv7]. Extra
```
Comment (wkalt) [inline documentation/api/command/v1/commands.markdown]: do we need a line on code_id here?

File: src/puppetlabs/puppetdb/catalogs.clj [hunk 2]
```diff
@@ -193,19 +189,22 @@
   [catalog]
   (utils/assoc-when catalog
                     :transaction_uuid nil
-                    :environment nil
-                    :api_version 1))
-
-(pls/defn-validated canonical-catalog
-  "Converts `catalog` to `version` in the canonical format, adding
-   and removing keys as needed"
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/catalogs.clj]: always hated this function :)

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 3]
```diff
@@ -342,15 +342,18 @@
...
-   :api_version api_version
    :catalog_version  version
    :transaction_uuid (sutils/munge-uuid-for-storage transaction_uuid)
    :timestamp (to-timestamp received-timestamp)
+   :code_id code_id
    :environment_id (ensure-environment environment)
-   :producer_timestamp (to-timestamp producer_timestamp)})
+   :producer_timestamp (to-timestamp producer_timestamp)
+   ;; FIXME THIS `api_version` IS NO LONGER USED IN ANY CAPACITY AND NEEDS TO BE
+   ;; REMOVED FROM OUR SCHEMA
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: this should be a ticket, not a comment. I think the ticket already exists.

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 4]
```diff
@@ -1350,6 +1350,9 @@
    "CREATE INDEX idx_reports_producer_timestamp
       ON reports(producer_timestamp)"))
 
+(defn add-code-id-to-catalogs []
+  (jdbc/do-commands "ALTER TABLE catalogs ADD code_id VARCHAR(255)"))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: how did we come to varchar(255)?

File: test/puppetlabs/puppetdb/examples.clj [hunk 5]
```diff
@@ -4,10 +4,10 @@
 (def catalogs
   {:empty
    {:certname         "empty.catalogs.com"
-    :api_version      1
     :version          "1330463884"
     :transaction_uuid "aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa"
     :environment      nil
+    :code_id nil
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/examples.clj]: should we have an actual string somewhere?

File: test/puppetlabs/puppetdb/scf/hash_test.clj [hunk 6]
```diff
@@ -190,20 +190,15 @@
...
           hash               (catalog-similarity-hash catalog)
-          ;; Functions that tweak various attributes of a catalog
-          tweak-api-version  #(update-in % [:api_version] inc)
-          tweak-version      #(update-in % [:version] str)
-          ;; List of all the tweaking functions
-          chaos-monkeys      [tweak-api-version tweak-version]
-          ;; Function that will apply a random tweak function
-          apply-monkey       #((rand-nth chaos-monkeys) %)]
+          chaos-monkey #(update-in % [:version] str)
+          ]
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/scf/hash_test.clj]: hanging

File: test/puppetlabs/puppetdb/examples.clj [hunk 7]
```diff
@@ -131,25 +132,29 @@
...
+          (update :resources conj {:type "File"
+                                   :title "/etc/foobar"
+                                   :exported false
+                                   :file "/tmp/foo"
+                                   :line 10
+                                   :tags ["file" "class" "foobar"]
+                                   :parameters {:ensure "directory"
+                                                :group  "root"
+                                                :user   "root"}}))}
+   })
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/examples.clj]: hanging

**Comment by wkalt** [comment]
+1 after minor comments

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 4 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: Did you have something else in mind for a semi-arbitrary string?

File: test/puppetlabs/puppetdb/examples.clj (hunk 5 above)
Comment (ajroetker) [inline test/puppetlabs/puppetdb/examples.clj]: I have some tests for that.

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 4 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: @senior Did you have any strong opinions on varchar(255) vs text?

**Comment by ajroetker** [comment]
@wkalt fixed up the hanging parens

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 4 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: as I understand it with postgres there's no upside to varchar over text here if we have no reason to cap the length. They'll be stored the same way, except varchar will involve some minimal-overhead bounds checking we don't need. text saves us a migration in the off chance they decide to use this for a long string.

File: src/puppetlabs/puppetdb/catalogs.clj [hunk 8]
```diff
@@ -193,19 +189,22 @@
...
+  (assoc catalog :code_id nil))
+
+(defn wire-v5->wire-v7 [catalog]
+  (-> catalog
+      (set/rename-keys {:name :certname})
+      utils/dash->underscore-keys
+      (dissoc :api_version)
+      wire-v6->wire-v7))
+
+(defn wire-v4->wire-v7 [catalog recieved-time]
```
Comment (mullr) [inline src/puppetlabs/puppetdb/catalogs.clj]: recieved -> received

File: src/puppetlabs/puppetdb/catalogs.clj [hunk 9]
```diff
@@ -332,8 +326,10 @@
     catalog))
 
 (def validate
-  "Function for validating v5- of the catalogs"
-  (comp validate-edges validate-resources #(s/validate (catalog-wireformat :all) %)))
+  "Function for validating v7- of the catalogs"
+  (comp validate-edges
+        validate-resources
+        #(s/validate catalog-wireformat-schema %)))
```
Comment (mullr) [inline src/puppetlabs/puppetdb/catalogs.clj]: `(partial s/validate catlaog-wireformat-schema)`

**Comment by mullr** [comment]
good to merge after the spelling is fixed :)

**Comment by mullr** [comment]
@ajroetker ⬆️

**Pull request closed without merging**